c++/src/capnp/compat/http-over-capnp.c++

```diff
@@ -688,6 +825,80 @@ public:
     });
   }
 
+  kj::Promise<void> connect(ConnectContext context) override {
+    auto params = context.getParams();
+    auto host = params.getHost();
+    auto headers = factory.headersToKj(params.getHeaders());
+    auto upPipe = kj::newOneWayPipe();
+
+    PipelineBuilder<ConnectResults> pb;
+    pb.setUp(factory.streamFactory.kjToCapnp(kj::mv(upPipe.out)));
+
+    context.setPipeline(pb.build());
+
+    // This is similar to kj::HttpClientAdapter::connect.
+    auto pipe = kj::newTwoWayPipe();
+    auto paf = kj::newPromiseAndFulfiller<kj::HttpClient::ConnectRequest::Status>();
+    auto response = kj::refcounted<HttpOverCapnpConnectResponseImpl>(kj::mv(paf.fulfiller));
+
+    auto promise = inner->connect(host, headers, *pipe.ends[0], *response)
+        .eagerlyEvaluate([response=kj::mv(response),
+                          host=kj::mv(host),
+                          headers=kj::mv(headers),
+                          connectStream=kj::mv(pipe.ends[0])](kj::Exception&& ex) mutable {
+        response->handleException(kj::mv(ex), kj::mv(connectStream));
+      });
+
+    return paf.promise.then(
```

Comment:
You should really be doing `return promise` here. Exceptions thrown by the service promise need to propagate out of the RPC. Currently you are just attaching the promise to another promise, which means exceptions will go nowhere, and in fact trailing non-promise-returning `.then()`s won't actually execute. This could be causing some of your hang problems.

I would get rid of `paf`, and move the code below that handles pumping the response to happen directly inside `HttpOverCapnpConnectResponseImpl`. Follow the model used in `HttpServiceResponseImpl`; this should be very similar. (The success case should look similar to the WebSocket case, which involves pumps in both directions.)

Reply:
Oh, I missed the `eagerlyEvaluate` above, which I suppose addresses the error handling and lazy evaluation issues.

It still feels weird that this is designed differently from the HTTP case... but I can't think of a specific reason it wouldn't work I guess.